Gems/Atom/Tools/AtomToolsFramework/Code/Source/Window/AtomToolsMainWindow.cpp

```diff
@@ -0,0 +1,208 @@
+        m_tabWidget->setObjectName("TabWidget");
+        m_tabWidget->setSizePolicy(QSizePolicy::Ignored, QSizePolicy::Preferred);
+        m_tabWidget->setContentsMargins(0, 0, 0, 0);
+
+        AtomToolsMainWindowRequestBus::Handler::BusConnect();
+    }
+
+    AtomToolsMainWindow::~AtomToolsMainWindow()
+    {
+        AtomToolsMainWindowRequestBus::Handler::BusDisconnect();
+    }
+
+    void AtomToolsMainWindow::ActivateWindow()
+    {
+        activateWindow();
+        raise();
+    }
+
+    bool AtomToolsMainWindow::AddDockWidget(const AZStd::string& name, QWidget* widget, uint32_t area, uint32_t orientation)
+    {
+        auto dockWidgetItr = m_dockWidgets.find(name);
+        if (dockWidgetItr != m_dockWidgets.end() || !widget)
+        {
+            return false;
+        }
+
+        auto dockWidget = new AzQtComponents::StyledDockWidget(name.c_str());
+        dockWidget->setObjectName(QString("%1_DockWidget").arg(name.c_str()));
+        dockWidget->setFeatures(QDockWidget::DockWidgetClosable | QDockWidget::DockWidgetFloatable | QDockWidget::DockWidgetMovable);
+        widget->setObjectName(name.c_str());
+        widget->setParent(dockWidget);
+        widget->setMinimumSize(QSize(300, 300));
+        dockWidget->setWidget(widget);
+        addDockWidget(aznumeric_cast<Qt::DockWidgetArea>(area), dockWidget);
+        resizeDocks({ dockWidget }, { 400 }, aznumeric_cast<Qt::Orientation>(orientation));
+        m_dockWidgets[name] = dockWidget;
+        return true;
+    }
+
+    void AtomToolsMainWindow::RemoveDockWidget(const AZStd::string& name)
+    {
+        auto dockWidgetItr = m_dockWidgets.find(name);
+        if (dockWidgetItr != m_dockWidgets.end())
+        {
+            delete dockWidgetItr->second;
+            m_dockWidgets.erase(dockWidgetItr);
+        }
+    }
+
+    void AtomToolsMainWindow::SetDockWidgetVisible(const AZStd::string& name, bool visible)
+    {
+        auto dockWidgetItr = m_dockWidgets.find(name);
+        if (dockWidgetItr != m_dockWidgets.end())
+        {
+            dockWidgetItr->second->setVisible(visible);
+        }
+    }
+
+    bool AtomToolsMainWindow::IsDockWidgetVisible(const AZStd::string& name) const
+    {
+        auto dockWidgetItr = m_dockWidgets.find(name);
+        if (dockWidgetItr != m_dockWidgets.end())
+        {
+            return dockWidgetItr->second->isVisible();
+        }
+        return false;
+    }
+
+    AZStd::vector<AZStd::string> AtomToolsMainWindow::GetDockWidgetNames() const
+    {
+        AZStd::vector<AZStd::string> names;
+        names.reserve(m_dockWidgets.size());
+        for (const auto& dockWidgetPair : m_dockWidgets)
+        {
+            names.push_back(dockWidgetPair.first);
+        }
+        return names;
+    }
+
+    void AtomToolsMainWindow::SetupMenu()
+    {
+        // Generating the main menu manually because it's easier and we will have some dynamic or data driven entries
```

Comment:
I'd prefer making `SetupMenu` a pure virtual here and having the child windows setup the menu

Reply:
Alternatively, you could move creation of the the bar and menu bar to their respective functions, maybe rename them to CreateMenu/CreateTabBar then have the derived class override to populate them.